test/functional/test_framework/mininode.py

```diff
@@ -440,3 +454,127 @@ def network_thread_join(timeout=10):
+                    self.send_message(msg_block(self.block_store[inv.hash]))
+
+    def on_getheaders(self, message):
+        """Search back through our block store for the locator, and reply with a headers message if found."""
+
+        locator, hash_stop = message.locator, message.hashstop
+
+        # Assume that the most recent block added is the tip
+        if len(self.block_store) == 0:
+            return
+        current_block_header = self.block_store[next(reversed(self.block_store))]
+        if current_block_header is None:
+            return
+
+        response = msg_headers()
+        headersList = [current_block_header]
+        maxheaders = 2000
+        while (headersList[0].sha256 not in locator.vHave):
+            # Walk back through the block store, adding headers to headersList
+            # as we go.
+            prevBlockHash = headersList[0].hashPrevBlock
+            if prevBlockHash in self.block_store:
+                prevBlockHeader = self.block_store[prevBlockHash]
+                headersList.insert(0, prevBlockHeader)
+            else:
+                break
+
+        # Truncate the list if there are too many headers
+        headersList = headersList[:maxheaders]
+        hashList = [x.sha256 for x in headersList]
+
+        # Stop the list at the hash_stop header if found
+        index = len(headersList)
+        if (hash_stop in hashList):
+            index = hashList.index(hash_stop) + 1
+        response.headers = headersList[:index]
+
+        if response is not None:
+            self.send_message(response)
+
+    def on_reject(self, message):
+        """Store reject reason and code for testing."""
+        self.reject_code_received = message.code
+        self.reject_reason_received = message.reason
+
+    def send_blocks_and_test(self, blocks, rpc, success=True, reject_code=None, reject_reason=None, request_block=True):
+        """Send blocks to test node and test whether the tip advances.
+
+         - add all blocks to our block_store
+         - send a headers message for the final block
+         - the on_getheaders handler will ensure that any getheaders are responded to
+         - if request_block is True: wait for getdata for each of the blocks. The on_getdata handler will
+           ensure that any getdata messages are responded to
+         - if reject_code and reject_reason are set: assert that the correct reject message is received
+         - if success is True: assert that the node's tip advances to the most recent block.
+         - if success is False: assert that the node's tip doesn't advance."""
+
+        with mininode_lock:
+            self.reject_code_received = None
+            self.reject_reason_received = None
+
+        for block in blocks:
+            self.block_store[block.sha256] = block
+
+        self.send_message(msg_headers([blocks[-1]]))
+
+        if request_block:
+            wait_until(lambda: blocks[-1].sha256 in self.getdata_requests)
```

Comment:
Since `getdata_requests` is written to with the mininode lock, why is a `lock=mininode_lock` not required when reading?

Reply:
You're right. This is required. Updating.